File: ios/Classes/SipSdkFlutterPlugin.swift
import Flutter
import SIPFramework
import UIKit

public class SipSdkFlutterPlugin: NSObject, FlutterPlugin {
    static var channel: FlutterMethodChannel?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "sip_sdk_flutter", binaryMessenger: registrar.messenger())
        let instance = SipSdkFlutterPlugin()
        SipSdkFlutterPlugin.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initSDK":
            if let args = call.arguments as? [String: Any] {
                initSDK(args: args, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Expected dictionary", details: nil))
            }
        case "registrar":
            if let args = call.arguments as? [String: Any] {
                registrar(args: args, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Expected dictionary", details: nil))
            }
        case "unRegistrar":
            if let args = call.arguments as? [String: Any] {
                unRegistrar(args: args, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Expected dictionary", details: nil))
            }
        case "call":
            if let args = call.arguments as? [String: Any] {
                self.call(args: args, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Expected dictionary", details: nil))
            }
        case "callIP":
            if let args = call.arguments as? [String: Any] {
                callIP(args: args, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Expected dictionary", details: nil))
            }
        case "answer":
            if let args = call.arguments as? [String: Any] {
                answer(args: args, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Expected dictionary", details: nil))
            }
        case "sendDtmfInfo":
            if let args = call.arguments as? [String: Any] {
                sendDtmfInfo(args: args, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Expected dictionary", details: nil))
            }
        case "sendMessage":
            if let args = call.arguments as? [String: Any] {
                sendMessage(args: args, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Expected dictionary", details: nil))
            }
        case "sendMessageIP":
            if let args = call.arguments as? [String: Any] {
                sendMessageIP(args: args, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Expected dictionary", details: nil))
            }
        case "hangup":
            if let args = call.arguments as? [String: Any] {
                hangup(args: args, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Expected dictionary", details: nil))
            }
        case "dump":
            if let args = call.arguments as? [String: Any] {
                dump(args: args, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Expected dictionary", details: nil))
            }
        case "handleIpChange":
            if let args = call.arguments as? [String: Any] {
                handleIpChange(args: args, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Expected dictionary", details: nil))
            }
        case "destroy":
            if let args = call.arguments as? [String: Any] {
                destroy(args: args, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Expected dictionary", details: nil))
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initSDK(args: [String: Any], result: @escaping FlutterResult) {
        // 1. 提取 stunConfig
        var stun: STUNConfig? = nil
        if let stunDict = args["stunConfig"] as? [String: Any] {
            let servers = (stunDict["servers"] as? [String]) ?? []
            let enableIPv6 = (stunDict["enableIPv6"] as? Bool) ?? false
            stun = STUNConfig(servers: servers, enableIPv6: enableIPv6)
        }

        // 2. 提取 mediaConfig（可选，如果你有用到）
        var mediaConfig: SIPSDKMediaConfig? = nil
        if let mediaDict = args["mediaConfig"] as? [String: Any] {
            mediaConfig = SIPSDKMediaConfig(
                audioClockRate: Int32(mediaDict["audioClockRate"] as? Int ?? 16000),
                micGain: (mediaDict["micGain"] as? NSNumber)?.floatValue ?? 1.0,
                speakerGain: (mediaDict["speakerGain"] as? NSNumber)?.floatValue ?? 1.0,
                nsEnable: (mediaDict["nsEnable"] as? Bool) ?? true,
                agcEnable: (mediaDict["agcEnable"] as? Bool) ?? true,
                aecEnable: (mediaDict["aecEnable"] as? Bool) ?? true,
                aecEliminationTime: Int16(mediaDict["aecEliminationTime"] as? Int ?? 30),
                notEnableEncode: (mediaDict["notEnableEncode"] as? Bool) ?? false,
                notEnableDecode: (mediaDict["notEnableDecode"] as? Bool) ?? false,
                decodeMaxWidth: UInt32(mediaDict["decodeMaxWidth"] as? Int ?? 1920),
                decodeMaxHeight: UInt32(mediaDict["decodeMaxHeight"] as? Int ?? 1080),
                combinSpsPpsIdr: (mediaDict["combinSpsPpsIdr"] as? Bool) ?? false,
                profileLevelId: mediaDict["profileLevelId"] as? String,
                packetizationMode: mediaDict["packetizationMode"] as? String
            )
        } else {
            mediaConfig = SIPSDKMediaConfig()
        }

        // 配置回调
        let callbacks = SIPSDKCallbacks(
            onLogCallback: SIPManage.onLogCallback,
            onInitCompleted: SIPManage.onInitCompleted,
            onStopCompleted: SIPManage.onStopCompleted,
            onRegistrarState: SIPManage.onRegistrarState,
            onIncomingCall: SIPManage.onIncomingCall,
            onDtmfInfo: SIPManage.onDtmfInfo,
            onMessage: SIPManage.onMessage,
            onMessageState: SIPManage.onMessageState,
            onCallState: SIPManage.onCallState,
            onExpireWarning: SIPManage.onExpireWarning
        )

        // 3. 提取 SIPSDKConfig 主结构体字段
        let config = SIPSDKConfig(
            port: UInt32(args["port"] as? Int ?? 58581),
            publicAddr: args["publicAddr"] as? String,
            logLevel: Int32(args["logLevel"] as? Int ?? 4),
            userAgent: args["userAgent"] as? String ?? "",
            workerThreadCount: Int32(args["workerThreadCount"] as? Int ?? 1),
            enableVideo: (args["enableVideo"] as? Bool) ?? true,
            sdkObserver: callbacks,
            allowMultipleConnections: (args["allowMultipleConnections"] as? Bool) ?? false,
            domainNameDirectRegistrar: (args["domainNameDirectRegistrar"] as? Bool) ?? false,
            doesItSupportBroadcast: (args["doesItSupportBroadcast"] as? Bool) ?? false,
            stunConfig: stun
        )
        let baseUrl: String = args["baseUrl"] as? String ?? ""
        let clientId: String = args["clientId"] as? String ?? ""
        let clientSecret: String = args["clientSecret"] as? String ?? ""
        SIPHandle.initSDK(baseUrl: baseUrl,
                          clientId: clientId,
                          clientSecret: clientSecret,
                          config: config,
                          mediaConfig: mediaConfig!)
        result(nil) // 表示成功
    }

    private func registrar(args: [String: Any], result: @escaping FlutterResult) {
        var localConfig: REGLocalConfig?
        if let localConfigDict = args["localConfig"] as? [String: Any] {
            let username = localConfigDict["username"] as? String
            let proxy = localConfigDict["proxy"] as? String
            let proxyPort = UInt32(localConfigDict["proxyPort"] as? Int ?? 0)
            let enableStreamControl = (localConfigDict["enableStreamControl"] as? Bool) ?? false
            let streamElapsed = Int32(localConfigDict["streamElapsed"] as? Int ?? 0)
            let startKeyframeCount = UInt32(localConfigDict["startKeyframeCount"] as? Int ?? 120)
            let startKeyframeInterval = UInt32(localConfigDict["startKeyframeInterval"] as? Int ?? 1000)

            localConfig = REGLocalConfig(
                username: username,
                proxy: proxy,
                proxyPort: proxyPort,
                enableStreamControl: enableStreamControl,
                streamElapsed: streamElapsed,
                startKeyframeCount: startKeyframeCount,
                startKeyframeInterval: startKeyframeInterval
            )
        }

        var turnConfig: TURNConfig?
        if let turnConfigDict = args["turnConfig"] as? [String: Any] {
            let enable = (turnConfigDict["enable"] as? Bool) ?? false
            let server = turnConfigDict["server"] as? String
            let realm = turnConfigDict["realm"] as? String
            let username = turnConfigDict["username"] as? String
            let password = turnConfigDict["password"] as? String

            turnConfig = TURNConfig(
                enable: enable,
                server: server,
                realm: realm,
                username: username,
                password: password
            )
        }

        let headers = (args["headers"] as? [String: String])?.map { ($0.key, $0.value) }

        let domain = args["domain"] as? String
        let username = args["username"] as? String
        let password = args["password"] as? String
        let transport = args["transport"] as? String
        let serverAddr = args["serverAddr"] as? String
        let serverPort = UInt32(args["serverPort"] as? Int ?? 5060)
        let proxy = args["proxy"] as? String
        let proxyPort = UInt32(args["proxyPort"] as? Int ?? 0)
        let enableStreamControl = (args["enableStreamControl"] as? Bool) ?? false
        let streamElapsed = Int32(args["streamElapsed"] as? Int ?? 0)
        let startKeyframeCount = UInt32(args["startKeyframeCount"] as? Int ?? 120)
        let startKeyframeInterval = UInt32(args["startKeyframeInterval"] as? Int ?? 1000)

        let config = REGConfig(
            domain: domain,
            username: username,
            password: password,
            transport: transport,
            serverAddr: serverAddr,
            serverPort: serverPort,
            headers: headers,
            proxy: proxy,
            proxyPort: proxyPort,
            enableStreamControl: enableStreamControl,
            streamElapsed: streamElapsed,
            startKeyframeCount: startKeyframeCount,
            startKeyframeInterval: startKeyframeInterval,
            turnConfig: turnConfig
        )

        // 这里调用你自己的注册接口，确保 localConfig 不为空时才调用
        if let localConfig = localConfig {
            SIPHandle.registrar(localConfig: localConfig, config: config)
        } else {
            print("localConfig is nil, cannot registrar")
        }
        result(nil) // 表示成功
    }

    /**
     * 解除注册到服务器
     */
    private func unRegistrar(args _: [String: Any], result: @escaping FlutterResult) {
        SIPHandle.unRegistrar()
        result(nil)
    }

    /**
     * 通过服务器呼叫
     * username: 对方用户名
     * headers: 自定义头信息
     */
    private func call(args: [String: Any], result: @escaping FlutterResult) {
        let username = args["username"] as? String
        let headers = (args["headers"] as? [String: String])?.map { ($0.key, $0.value) }
        let sipCallParam = SIPCallParam(direction: 1,
                                        callType: SDK_MESSAGE_TYPE_SERVER.rawValue,
                                        username: username,
                                        headers: headers)
        DispatchQueue.main.async {
            let callVC = CallViewController(callParam: sipCallParam)
            NavigationCoordinator.shared.present(callVC)
        }
        result(nil)
    }

    /**
     * 通过服IP呼叫
     * ip: 对方IP
     * headers: 自定义头信息
     */
    private func callIP(args: [String: Any], result: @escaping FlutterResult) {
        let ip = args["ip"] as? String
        let headers = (args["headers"] as? [String: String])?.map { ($0.key, $0.value) }
        let sipCallParam = SIPCallParam(direction: 1,
                                        callType: SDK_MESSAGE_TYPE_IP.rawValue,
                                        remoteIP: ip,
                                        headers: headers)
        DispatchQueue.main.async {
            let callVC = CallViewController(callParam: sipCallParam)
            NavigationCoordinator.shared.present(callVC)
        }
        result(nil)
    }

    /**
     * 接听呼叫
     * code: 接听状态码，正常接听200，先通媒体183
     * callUuid: 接听所有0，接听指定呼叫不等于0
     */
    private func answer(args: [String: Any], result: @escaping FlutterResult) {
        let code = UInt32(args["code"] as? Int ?? 200)
        let callUUID = UInt(args["callUUID"] as? Int ?? 0)
        SIPHandle.answer(code: code, callUuid: callUUID)
        result(nil)
    }

    /**
     * 发送info消息
     * type: 消息类型
     * contentType: 内容类型
     * content: 内容（除自定义类型外、其他的类型内容只能是一个字节）
     * callUuid: 为0所有呼叫发送，不等于0指定呼叫发送
     */
    private func sendDtmfInfo(args: [String: Any], result: @escaping FlutterResult) {
        let dtmfInfoType = Int32(args["dtmfInfoType"] as? Int ?? 0)
        let content = args["content"] as? String ?? ""
        let callUUID = UInt(args["callUUID"] as? Int ?? 0)
        // 发送
        SIPHandle.sendDtmfInfo(type: dtmfInfoType, callUuid: callUUID, content: content)
        // 成功回调
        result(nil)
    }

    /**
     * 通过服务器发送sip message消息
     * username: 对方账号
     * content: 内容
     */
    private func sendMessage(args: [String: Any], result: @escaping FlutterResult) {
        let username = args["username"] as? String ?? ""
        let content = args["content"] as? String ?? ""
        // 发送
        SIPHandle.sendMessage(username: username, content: content)
        // 成功回调
        result(nil)
    }

    /**
     * 通过IP发送sip message消息
     * username: 对方账号
     * content: 内容
     */
    private func sendMessageIP(args: [String: Any], result: @escaping FlutterResult) {
        let ip = args["ip"] as? String ?? ""
        let content = args["content"] as? String ?? ""
        // 发送
        SIPHandle.sendMessageIP(ip: ip, content: content)
        // 成功回调
        result(nil)
    }

    /**
     * 挂断 call_uuid 对应的呼叫
     * code: 挂断状态码，正常挂断200
     * callUuid: 为0挂断所有呼叫，不等于0挂断指定呼叫
     */
    private func hangup(args: [String: Any], result: @escaping FlutterResult) {
        let code = UInt32(args["code"] as? Int ?? 200)
        let callUUID = UInt(args["callUUID"] as? Int ?? 0)
        SIPHandle.hangup(code: code, callUuid: callUUID)
        result(nil)
    }

    /**
     *  打印SDK信息，包括所有内存使用信息
     */
    private func dump(args _: [String: Any], result: @escaping FlutterResult) {
        SIPHandle.dump()
        result(nil)
    }

    /**
     *  IP 发生改变调用
     */
    private func handleIpChange(args _: [String: Any], result: @escaping FlutterResult) {
        SIPHandle.handleIpChange()
        result(nil)
    }

    /**
     *  销毁
     */
    private func destroy(args _: [String: Any], result: @escaping FlutterResult) {
        SIPHandle.destroy()
        result(nil)
    }
}

File: android/src/main/java/com/sip/flutter/sip_sdk_flutter/SipSdkFlutterPlugin.java
package com.sip.flutter.sip_sdk_flutter;

import static com.sip.sdk.entity.SDKConstants.SDK_CALL_TYPE_IP;
import static com.sip.sdk.entity.SDKConstants.SDK_CALL_TYPE_SERVER;
import static com.sip.sdk.entity.SDKConstants.SDK_DTMF_INFO_TYPE;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;

import androidx.annotation.NonNull;

import com.sip.flutter.sip_sdk_flutter.activity.CallActivity;
import com.sip.flutter.sip_sdk_flutter.activity.entity.CMSIPSDKCallParam;
import com.sip.flutter.sip_sdk_flutter.sip.SIPManage;
import com.sip.flutter.sip_sdk_flutter.utils.MapUtils;
import com.sip.sdk.SIPSDK;
import com.sip.sdk.entity.SIPSDKConfig;
import com.sip.sdk.entity.SIPSDKLocalConfig;
import com.sip.sdk.entity.SIPSDKMediaConfig;
import com.sip.sdk.entity.SIPSDKMediaH264Fmtp;
import com.sip.sdk.entity.SIPSDKRegistrarConfig;
import com.sip.sdk.entity.SIPSDKStunConfig;
import com.sip.sdk.entity.SIPSDKTurnConfig;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.Log;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/**
 * SipSdkFlutterPlugin
 */
public class SipSdkFlutterPlugin implements FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    public static MethodChannel channel;

    @SuppressLint("StaticFieldLeak")
    public static Context context;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "sip_sdk_flutter");
        channel.setMethodCallHandler(this);
        SipSdkFlutterPlugin.context = flutterPluginBinding.getApplicationContext();
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (call.method.equals("initSDK")) {
            initSDK(call.arguments(), result);
        } else if (call.method.equals("registrar")) {
            registrar(call.arguments(), result);
        } else if (call.method.equals("unRegistrar")) {
            unRegistrar(call.arguments(), result);
        } else if (call.method.equals("call")) {
            call(call.arguments(), result);
        } else if (call.method.equals("callIP")) {
            callIP(call.arguments(), result);
        } else if (call.method.equals("answer")) {
            answer(call.arguments(), result);
        } else if (call.method.equals("sendDtmfInfo")) {
            sendDtmfInfo(call.arguments(), result);
        } else if (call.method.equals("sendMessage")) {
            sendMessage(call.arguments(), result);
        } else if (call.method.equals("sendMessageIP")) {
            sendMessageIP(call.arguments(), result);
        } else if (call.method.equals("hangup")) {
            hangup(call.arguments(), result);
        } else if (call.method.equals("dump")) {
            dump(call.arguments(), result);
        } else if (call.method.equals("destroy")) {
            destroy(call.arguments(), result);
        } else if (call.method.equals("handleIpChange")) {
            handleIpChange(call.arguments(), result);
        } else {
            result.notImplemented();
        }
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }

    private void initSDK(Map<String, Object> args, MethodChannel.Result result) {
        Map<String, Object> stunDict = MapUtils.getMap(args, "stunConfig");
        SIPSDKStunConfig stunConfig = null;
        if (stunDict != null) {
            List<String> servers = MapUtils.get(stunDict, "servers", new ArrayList<>());
            boolean enableIPv6 = MapUtils.get(stunDict, "enableIPv6", false);
            stunConfig = new SIPSDKStunConfig();
            stunConfig.count = servers.size();
            stunConfig.servers = servers;
            stunConfig.enableIpv6 = enableIPv6;
        }

        Map<String, Object> mediaDict = MapUtils.getMap(args, "mediaConfig");
        SIPSDKMediaConfig mediaConfig = new SIPSDKMediaConfig();
        if (mediaDict != null) {
            mediaConfig.audioClockRate = MapUtils.get(mediaDict, "audioClockRate", 16000);
            mediaConfig.micGain = MapUtils.get(mediaDict, "micGain", 1.0f);
            mediaConfig.speakerGain = MapUtils.get(mediaDict, "speakerGain", 1.0f);
            mediaConfig.nsEnable = MapUtils.get(mediaDict, "nsEnable", true);
            mediaConfig.agcEnable = MapUtils.get(mediaDict, "agcEnable", true);
            mediaConfig.aecEnable = MapUtils.get(mediaDict, "aecEnable", true);
            mediaConfig.aecEliminationTime = MapUtils.get(mediaDict, "aecEliminationTime", (short) 30);
            mediaConfig.notEnableEncode = MapUtils.get(mediaDict, "notEnableEncode", false);
            mediaConfig.notEnableDecode = MapUtils.get(mediaDict, "notEnableDecode", false);
            mediaConfig.decodeMaxWidth = MapUtils.get(mediaDict, "decodeMaxWidth", 1920);
            mediaConfig.decodeMaxHeight = MapUtils.get(mediaDict, "decodeMaxHeight", 1080);
            mediaConfig.combinSpsPpsIdr = MapUtils.get(mediaDict, "combinSpsPpsIdr", false);
            String profileLevelId = MapUtils.get(mediaDict, "profileLevelId", null);
            String packetizationMode = MapUtils.get(mediaDict, "packetizationMode", null);
            if (profileLevelId != null && !profileLevelId.isEmpty() && packetizationMode != null && !packetizationMode.isEmpty()) {
                if (mediaConfig.h264Fmtp == null) {
                    mediaConfig.h264Fmtp = new SIPSDKMediaH264Fmtp();
                }
                mediaConfig.h264Fmtp.profileLevelId = profileLevelId;
                mediaConfig.h264Fmtp.packetizationMode = packetizationMode;
            }
        }

        SIPSDKConfig config = new SIPSDKConfig();
        config.port = MapUtils.get(args, "port", 58581);
        config.logLevel = MapUtils.get(args, "logLevel", 4);
        config.userAgent = MapUtils.get(args, "userAgent", "");
        config.workerThreadCount = MapUtils.get(args, "workerThreadCount", 1);
        config.videoEnable = MapUtils.get(args, "enableVideo", true);
        config.videoOutAutoTransmit = MapUtils.get(args, "videoOutAutoTransmit", true);
        config.allowMultipleConnections = MapUtils.get(args, "allowMultipleConnections", false);
        config.domainNameDirectRegistrar = MapUtils.get(args, "domainNameDirectRegistrar", false);
        config.doesItSupportBroadcast = MapUtils.get(args, "doesItSupportBroadcast", false);
        config.stunConfig = stunConfig;
        String baseUrl = MapUtils.get(args, "baseUrl", "");
        String clientId = MapUtils.get(args, "clientId", "");
        String clientSecret = MapUtils.get(args, "clientSecret", "");
        SIPManage.instance().init(context, baseUrl, clientId, clientSecret, config, mediaConfig);
        result.success(null);
    }

    private void registrar(Map<String, Object> args, MethodChannel.Result result) {
        Map<String, Object> localDict = MapUtils.getMap(args, "localConfig");
        SIPSDKLocalConfig localConfig = null;
        if (localDict != null) {
            localConfig = new SIPSDKLocalConfig();
            localConfig.username = MapUtils.get(localDict, "username", null);
            localConfig.proxy = MapUtils.get(localDict, "proxy", null);
            localConfig.proxyPort = MapUtils.get(localDict, "proxyPort", 0);
            localConfig.enableStreamControl = MapUtils.get(localDict, "enableStreamControl", false);
            localConfig.streamElapsed = MapUtils.get(localDict, "streamElapsed", 0);
            localConfig.startKeyframeCount = MapUtils.get(localDict, "startKeyframeCount", 120);
            localConfig.startKeyframeInterval = MapUtils.get(localDict, "startKeyframeInterval", 1000);
        }

        Map<String, Object> turnDict = MapUtils.getMap(args, "turnConfig");
        SIPSDKTurnConfig turnConfig = null;
        if (turnDict != null) {
            turnConfig = new SIPSDKTurnConfig();
            turnConfig.enable = MapUtils.get(turnDict, "enable", false);
            turnConfig.server = MapUtils.get(turnDict, "server", null);
            turnConfig.realm = MapUtils.get(turnDict, "realm", null);
            turnConfig.username = MapUtils.get(turnDict, "username", null);
            turnConfig.password = MapUtils.get(turnDict, "password", null);
        }

        Map<String, String> headers = new HashMap<>();
        Map<String, Object> rawHeaders = MapUtils.getMap(args, "headers");
        if (rawHeaders != null) {
            for (Map.Entry<String, Object> entry : rawHeaders.entrySet()) {
                if (entry.getValue() instanceof String) {
                    headers.put(entry.getKey(), (String) entry.getValue());
                }
            }
        }

        SIPSDKRegistrarConfig config = new SIPSDKRegistrarConfig();
        config.domain = MapUtils.get(args, "domain", null);
        config.username = MapUtils.get(args, "username", null);
        config.password = MapUtils.get(args, "password", null);
        config.transport = MapUtils.get(args, "transport", null);
        config.serverAddr = MapUtils.get(args, "serverAddr", null);
        config.serverPort = MapUtils.get(args, "serverPort", 5060);
        config.proxy = MapUtils.get(args, "proxy", null);
        config.proxyPort = MapUtils.get(args, "proxyPort", 5060);
        config.enableStreamControl = MapUtils.get(args, "enableStreamControl", false);
        config.streamElapsed = MapUtils.get(args, "streamElapsed", 0);
        config.startKeyframeCount = MapUtils.get(args, "startKeyframeCount", 120);
        config.startKeyframeInterval = MapUtils.get(args, "startKeyframeInterval", 1000);
        config.headers = headers;
        config.turnConfig = turnConfig;

        SIPSDK.registrarAccount(config);
        result.success(null);
    }

    /**
     * 解除注册到服务器
     */
    private void unRegistrar(Map<String, Object> args, MethodChannel.Result result) {
        SIPSDK.unRegistrar();
        result.success(null);
    }

    private void call(Map<String, Object> args, MethodChannel.Result result) {
        String username = MapUtils.get(args, "username", null);
        Map<String, String> headers = new HashMap<>();
        Map<String, Object> rawHeaders = MapUtils.getMap(args, "headers");
        if (rawHeaders != null) {
            for (Map.Entry<String, Object> entry : rawHeaders.entrySet()) {
                if (entry.getValue() instanceof String) {
                    headers.put(entry.getKey(), (String) entry.getValue());
                }
            }
        }

        CMSIPSDKCallParam callParam = new CMSIPSDKCallParam();
        callParam.direction = 1;
        callParam.callType = SDK_CALL_TYPE_SERVER;
        callParam.username = username;
        callParam.headers = new HashMap<>();

        Intent intent = new Intent(context, CallActivity.class);
        intent.putExtra("callParam", callParam);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(intent);

        result.success(null);
    }

    private void callIP(Map<String, Object> args, MethodChannel.Result result) {
        String ip = MapUtils.get(args, "ip", null);
        Map<String, String> headers = new HashMap<>();
        Map<String, Object> rawHeaders = MapUtils.getMap(args, "headers");
        if (rawHeaders != null) {
            for (Map.Entry<String, Object> entry : rawHeaders.entrySet()) {
                if (entry.getValue() instanceof String) {
                    headers.put(entry.getKey(), (String) entry.getValue());
                }
            }
        }
        CMSIPSDKCallParam callParam = new CMSIPSDKCallParam();
        callParam.direction = 1;
        callParam.callType = SDK_CALL_TYPE_IP;
        callParam.remoteIp = ip;
        callParam.headers = new HashMap<>();

        Intent intent = new Intent(context, CallActivity.class);
        intent.putExtra("callParam", callParam);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(intent);

        result.success(null);
    }

    private void answer(Map<String, Object> args, MethodChannel.Result result) {
        int code = MapUtils.get(args, "200", 200);
        long callUUID = MapUtils.get(args, "callUUID", 0);
        SIPSDK.answer(code, callUUID);
        result.success(null);
    }

    private void sendDtmfInfo(Map<String, Object> args, MethodChannel.Result result) {
        long callUUID = MapUtils.get(args, "callUUID", 0);
        int dtmfInfoType = MapUtils.get(args, "dtmfInfoType", SDK_DTMF_INFO_TYPE);
        String content = MapUtils.get(args, "content", null);
        String contentType = MapUtils.get(args, "contentType", null);
        SIPSDK.sendDtmfInfo(dtmfInfoType, contentType, content);
        result.success(null);
    }

    private void sendMessage(Map<String, Object> args, MethodChannel.Result result) {
        String username = MapUtils.get(args, "username", null);
        String content = MapUtils.get(args, "content", null);
        SIPSDK.sendMessage(username, content);
        result.success(null);
    }

    private void sendMessageIP(Map<String, Object> args, MethodChannel.Result result) {
        String ip = MapUtils.get(args, "ip", null);
        String content = MapUtils.get(args, "content", null);
        SIPSDK.sendMessageIP(ip, content);
        result.success(null);
    }

    private void hangup(Map<String, Object> args, MethodChannel.Result result) {
        int code = MapUtils.get(args, "code", 487);
        long callUUID = MapUtils.get(args, "callUUID", 0);
        if (callUUID == 0) {
            SIPSDK.hangup(code);
        } else {
            SIPSDK.hangupWithUuid(code, callUUID);
        }
        result.success(null);
    }

    /**
     * 打印SDK信息，包括所有内存使用信息
     */
    private void dump(Map<String, Object> args, MethodChannel.Result result) {
        SIPSDK.dump();
        result.success(null);
    }

    private void destroy(Map<String, Object> args, MethodChannel.Result result) {
        SIPSDK.destroy();
        result.success(null);
    }

    private void handleIpChange(Map<String, Object> args, MethodChannel.Result result) {
        SIPSDK.handleIpChange();
        result.success(null);
    }
}
